src/protocol/ipa_prf/prf_sharding/mod.rs

```diff
@@ -213,30 +196,42 @@ impl InputsRequiredFromPrevRow {
 }
 
 #[derive(Debug)]
-pub struct CappedAttributionOutputs {
-    pub attributed_breakdown_key_bits: BitDecomposed<Replicated<Gf2>>,
-    pub capped_attributed_trigger_value: BitDecomposed<Replicated<Gf2>>,
+pub struct CappedAttributionOutputs<
+    BK: WeakSharedValue + CustomArray<Element = Boolean>,
+    TV: WeakSharedValue + CustomArray<Element = Boolean>,
+> {
+    pub attributed_breakdown_key_bits: Replicated<BK>,
+    pub capped_attributed_trigger_value: Replicated<TV>,
 }
 
-impl ToBitConversionTriples for CappedAttributionOutputs {
+impl<
+        BK: WeakSharedValue + CustomArray<Element = Boolean>,
+        TV: WeakSharedValue + CustomArray<Element = Boolean>,
+    > ToBitConversionTriples for CappedAttributionOutputs<BK, TV>
+{
     type Residual = ();
 
     fn bits(&self) -> u32 {
-        (self.attributed_breakdown_key_bits.len() + self.capped_attributed_trigger_value.len())
-            .try_into()
-            .unwrap()
+        BK::BITS + TV::BITS
     }
 
     fn triple<F: PrimeField>(&self, role: Role, i: u32) -> BitConversionTriple<Replicated<F>> {
         assert!(i < self.bits());
         let i: usize = i.try_into().unwrap();
-        let bit = if i < self.attributed_breakdown_key_bits.len() {
-            &self.attributed_breakdown_key_bits[i]
+        let bk_bits: usize = BK::BITS.try_into().unwrap();
+        let bit_left = if i < bk_bits {
+            self.attributed_breakdown_key_bits.0.get(i).unwrap() == Boolean::ONE
         } else {
-            let i = i - self.attributed_breakdown_key_bits.len();
-            &self.capped_attributed_trigger_value[i]
+            let i = i - bk_bits;
+            self.capped_attributed_trigger_value.0.get(i).unwrap() == Boolean::ONE
         };
-        BitConversionTriple::new(role, bit.left() == Gf2::ONE, bit.right() == Gf2::ONE)
+        let bit_right = if i < bk_bits {
+            self.attributed_breakdown_key_bits.1.get(i).unwrap() == Boolean::ONE
```

Comment:
Isn't this cleaner?

```suggestion
            self.attributed_breakdown_key_bits.right()[i] == Boolean::ONE
```

Reply:
I agree it would be... but there is no indexing operation available.

Reply:
Maybe you can interpret the comment as a request to add `impl Index<usize> for $name` to the macro then.